polygeist/lib/Dialect/Polygeist/Transforms/FunctionSpecialization.cpp

```diff
@@ -0,0 +1,281 @@
+#include "mlir/Dialect/Polygeist/IR/Ops.h"
+#include "mlir/Dialect/Polygeist/IR/Polygeist.h"
+#include "mlir/Dialect/Polygeist/Utils/TransformUtils.h"
+#include "mlir/Dialect/SYCL/IR/SYCLOps.h"
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "function-specialization"
+
+namespace mlir {
+namespace polygeist {
+#define GEN_PASS_DEF_FUNCTIONSPECIALIZATION
+#include "mlir/Dialect/Polygeist/Transforms/Passes.h.inc"
+} // namespace polygeist
+} // namespace mlir
+
+using namespace mlir;
+using AccessorPtrType = VersionConditionBuilder::AccessorPtrType;
+using AccessorPtrPairType = VersionConditionBuilder::AccessorPtrPairType;
+
+static llvm::cl::opt<unsigned> functionSpecializationAccessorLimit(
+    "function-specialization-accessor-limit", llvm::cl::init(5),
+    llvm::cl::desc(
+        "Maximum number of accessors allowed for function specialization"));
+
+//===----------------------------------------------------------------------===//
+// Helper Functions
+//===----------------------------------------------------------------------===//
+
+/// Returns true if \p type is 'memref<?x!sycl.accessor>', and false otherwise.
+static bool isValidMemRefType(Type type) {
+  auto mt = dyn_cast<MemRefType>(type);
+  bool isMemRefWithExpectedShape =
+      (mt && mt.hasRank() && (mt.getRank() == 1) &&
+       ShapedType::isDynamic(mt.getShape()[0]) && mt.getLayout().isIdentity());
+  if (!isMemRefWithExpectedShape)
+    return false;
+
+  return isa<sycl::AccessorType>(mt.getElementType());
+}
+
+/// Returns true if \p arg is a candidate argument. Currently, all arguments
+/// with valid memref type determined by isValidMemRefType is a candidate
+/// argument, i.e., 'memref<?x!sycl.accessor>`.
+static bool isCandidateArg(Value arg) {
+  if (isValidMemRefType(arg.getType())) {
+    assert(isa<AccessorPtrType>(arg) &&
+           "Expecting valid memref type to be AccessorPtrType");
+    return true;
+  }
+  return false;
+}
+
+/// Populate \p candArgs with candidate arguments from \p call.
+static void
+collectCandidateArguments(CallOpInterface call,
+                          SmallVectorImpl<AccessorPtrType> &candArgs) {
+  assert(candArgs.empty() && "Expecting empty candArgs");
+  for (Value arg : call.getArgOperands())
+    if (isCandidateArg(arg))
+      candArgs.push_back(cast<AccessorPtrType>(arg));
+}
+
+/// Returns the cloned version of \p func.
+static FunctionOpInterface cloneFunction(FunctionOpInterface func) {
+  FunctionOpInterface clonedFunc = func.clone();
+  std::string newFnName = (func.getName() + ".specialized").str();
+#ifndef NDEBUG
+  ModuleOp module = func->getParentOfType<ModuleOp>();
+  module->walk([newFnName](FunctionOpInterface func) {
+    assert(func.getName() != newFnName &&
+           "Expecting new function name to be unique");
+  });
+#endif // NDEBUG
+  clonedFunc.setName(newFnName);
+  OpBuilder builder(func);
+  builder.insert(clonedFunc);
+  privatize(clonedFunc);
+  return clonedFunc;
+}
```

Comment:
```suggestion
  OpBuilder builder(func);
  FunctionOpInterface clonedFunc = builder.clone(func);
  clonedFunc.setName(newFnName);
  privatize(clonedFunc);
  return clonedFunc;
}
```
Clone "in-place"

Reply:
reordered, but if use `builder.clone(func)`, get the error below:
```
polygeist-opt: mlir/lib/IR/Operation.cpp:392: void llvm::ilist_traits<mlir::Operation>::addNodeToList(llvm::ilist_traits<mlir::Operation>::Operation *): Assertion `!op->getBlock() && "already in an operation block!"' failed.
```

Reply:
Oh, I see. Didn't know that'd happen. Thanks!